cmd/apiserver_receive_adapter/main.go

```diff
@@ -60,52 +66,84 @@ type envConfig struct {
 	Kind          StringList `required:"true"`
 	Controller    []bool     `required:"true"`
 	LabelSelector StringList `envconfig:"SELECTOR" required:"true"`
+	Name          string     `envconfig:"NAME" required:"true"`
+	// MetricsConfigBase64 is a base64 encoded json string of
+	// metrics.ExporterOptions. This is used to configure the metrics exporter
+	// options, the config is stored in a config map inside the controllers
+	// namespace and copied here.
+	MetricsConfigBase64 string `envconfig:"K_METRICS_CONFIG" required:"true"`
+
+	// LoggingConfigBase64 is a base64 encoded json string of logging.Config.
+	// This is used to configure the logging config, the config is stored in
+	// a config map inside the controllers namespace and copied here.
+	LoggingConfigBase64 string `envconfig:"K_LOGGING_CONFIG" required:"true"`
 }
 
 // TODO: the controller should take the list of GVR
 
 func main() {
 	flag.Parse()
 
-	logCfg := zap.NewProductionConfig()
-	logCfg.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
-	dlogger, err := logCfg.Build()
+	var env envConfig
+	err := envconfig.Process("", &env)
 	if err != nil {
-		log.Fatalf("Error building logger: %v", err)
+		panic(fmt.Sprintf("Error processing env var: %s", err))
+	}
+	// TODO move this util to pkg
+	// Convert base64 encoded json logging.Config to logging.Config.
+	loggingConfig, err := utils.Base64ToLoggingConfig(env.LoggingConfigBase64)
+	if err != nil {
+		fmt.Printf("[ERROR] failed to process logging config: %s", err.Error())
+		// Use default logging config.
+		if loggingConfig, err = logging.NewConfigFromMap(map[string]string{}); err != nil {
+			// If this fails, there is no recovering.
+			panic(err)
+		}
+	}
+	loggerSugared, _ := logging.NewLoggerFromConfig(loggingConfig, component)
+	logger := loggerSugared.Desugar()
+	defer flush(loggerSugared)
+
+	// Convert base64 encoded json metrics.ExporterOptions to
+	// metrics.ExporterOptions.
+	metricsConfig, err := utils.Base64ToMetricsOptions(
+		env.MetricsConfigBase64)
+	if err != nil {
+		logger.Error("failed to process metrics options ", zap.Error(err))
```

Comment:
```suggestion
		logger.Error("failed to process metrics options", zap.Error(err))
```

Don't need the extra space when using the desugared logger. Not here or the other lines.

Once, again, I'll let it submit as-is, please send a follow up PR.

Reply:
Will do, thanks @Harwayne!